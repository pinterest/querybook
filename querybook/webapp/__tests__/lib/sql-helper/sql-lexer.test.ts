import {
    findTableReferenceAndAlias,
    findWithStatementPlaceholder,
    getQueryAsExplain,
    getQueryKeywords,
    getQueryLinePosition,
    getStatementType,
    simpleParse,
    tokenize,
    tokenPatternMatch,
} from 'lib/sql-helper/sql-lexer';

const simpleQuery = `
SELECT
  aa.a,
  aa.b
FROM
  table_a aa
  JOIN table_b bb ON aa.a = bb.b
WHERE
  aa.b = 3
ORDER BY
  bb.b
LIMIT
  5
`;

test('tokenize', () => {
    expect(tokenize('select * from table where a = 5;')).toMatchSnapshot();
    expect(tokenize(simpleQuery)).toMatchSnapshot();
});

test('simpleParse', () => {
    expect(
        simpleParse(
            tokenize(`
    select * from table where fielda = 5;
    select * from table where fieldb = (select fielda from table where fieldb = 5);
    `)
        )
    ).toMatchSnapshot();

    expect(
        simpleParse(
            tokenize(`select * from table where extract(YEAR FROM fielda) = 5;`)
        )
    ).toMatchSnapshot();
});

test('findWithStatementPlaceholder', () => {
    // 0 case
    expect(
        findWithStatementPlaceholder(
            simpleParse(tokenize(`select * from z`))[0]
        )
    ).toEqual([]);

    // 1 case
    expect(
        findWithStatementPlaceholder(
            simpleParse(
                tokenize(
                    `
with table_3 as (
    SELECT * from table_4
)
SELECT *
FROM table_2
WHERE table_2.field_1 = (
    SELECT SUM(table_1.field_3)
    FROM table_1 JOIN table_3 ON table_1.field_3 = table_3.field_2
    WHERE table_1.field_1 = table_2.field_2
)`
                )
            )[0]
        )
    ).toEqual(['table_3']);

    // multi case
    expect(
        findWithStatementPlaceholder(
            simpleParse(
                tokenize(
                    `with p as (
    select *
    from (values
        ('a', 1, 1),
        ('b', 2, null),
        ('c', null, 3),
        ('d', null, null)
    ) t1 (letter, val1, val2)
    ), z as (
    select *, val1 IS DISTINCT FROM val2
    from p
    order by letter
    )
    select * from z`,
                    { language: 'presto' }
                )
            )[0]
        )
    ).toEqual(['p', 'z']);
});

test('findTableReferenceAndAlias', () => {
    const tokenTableA = {
        end: 9,
        line: 5,
        name: 'table_a',
        schema: 'default',
        start: 2,
    };
    const tokenTableB = {
        end: 14,
        line: 6,
        name: 'table_b',
        schema: 'default',
        start: 7,
    };

    expect(
        findTableReferenceAndAlias(simpleParse(tokenize(simpleQuery)))
    ).toEqual({
        aliases: {
            0: {
                aa: tokenTableA,
                bb: tokenTableB,
            },
        },
        references: { 0: [tokenTableA, tokenTableB] },
    });

    expect(
        findTableReferenceAndAlias(
            simpleParse(
                tokenize(
                    `
SELECT *
FROM table_2 t2
JOIN table_1 ON table_1.field_1 = t2.field_2
AND extract(YEAR FROM field_1_date) = t2.field_year`
                )
            )
        )
    ).toEqual({
        aliases: {
            0: {
                t2: {
                    end: 12,
                    line: 2,
                    name: 'table_2',
                    schema: 'default',
                    start: 5,
                },
            },
        },
        references: {
            0: [
                {
                    end: 12,
                    line: 2,
                    name: 'table_2',
                    schema: 'default',
                    start: 5,
                },
                {
                    end: 12,
                    line: 3,
                    name: 'table_1',
                    schema: 'default',
                    start: 5,
                },
            ],
        },
    });

    expect(
        findTableReferenceAndAlias(
            simpleParse(
                tokenize(
                    `
with table_3 as (
    SELECT * from table_4
)
SELECT *
FROM table_2 AS t2
WHERE table_2.field_1 = (
    SELECT SUM(table_1.field_3)
    FROM table_1 JOIN table_3 ON table_1.field_3 = table_3.field_2
    WHERE table_1.field_1 = table_2.field_2
)`
                )
            )
        )
    ).toEqual({
        aliases: {
            0: {
                t2: {
                    end: 12,
                    line: 5,
                    name: 'table_2',
                    schema: 'default',
                    start: 5,
                },
            },
        },
        references: {
            0: [
                {
                    end: 25,
                    line: 2,
                    name: 'table_4',
                    schema: 'default',
                    start: 18,
                },
                {
                    end: 12,
                    line: 5,
                    name: 'table_2',
                    schema: 'default',
                    start: 5,
                },
                {
                    end: 16,
                    line: 8,
                    name: 'table_1',
                    schema: 'default',
                    start: 9,
                },
            ],
        },
    });

    expect(
        findTableReferenceAndAlias(
            simpleParse(
                tokenize(`
        with p as (
            select *
            from (values
              ('a', 1, 1),
              ('b', 2, null),
              ('c', null, 3),
              ('d', null, null)
            ) t1 (letter, val1, val2)
          ), z as (
            select *, val1 IS DISTINCT FROM val2
            from p
            order by letter
          )
          select * from z
        `)
            )
        )
    ).toEqual({
        aliases: {
            0: {},
        },
        references: { 0: [] },
    });
});

test('getQueryAsExplain', () => {
    expect(getQueryAsExplain('select 1')).toBe('EXPLAIN select 1;');
    expect(
        getQueryAsExplain(
            `select 1; --ignore
select 2;
select 3 --test`
        )
    ).toBe(
        `EXPLAIN select 1;
EXPLAIN select 2;
EXPLAIN select 3;`
    );
});

test('getQueryLinePosition', () => {
    expect(
        getQueryLinePosition(
            `SELECT *
FROM
    table_a;`
        )
    ).toStrictEqual([0, 9, 14, 27]);
});

test('getQueryKeywords', () => {
    expect(getQueryKeywords(simpleQuery)).toEqual(['select']);

    // multiple statements
    expect(
        getQueryKeywords(`INSERT INTO abc SELECT * FROM foobar;
        WITH foo as (SELECT * FROM hello.world)
        CREATE TABLE egg.spam AS
        SELECT a, b FROM foo`)
    ).toEqual(['insert', 'create']);
});
describe('getStatementType', () => {
    test('Simple example query', () => {
        expect(getStatementType(simpleParse(tokenize(simpleQuery))[0])).toEqual(
            'select'
        );
    });

    test('Simple insert query', () => {
        expect(
            getStatementType(
                simpleParse(
                    tokenize(`INSERT INTO abc SELECT * FROM foobar

            `)
                )[0]
            )
        ).toEqual('insert');
    });

    test('Simple create with "with" statement', () => {
        expect(
            getStatementType(
                simpleParse(
                    tokenize(`WITH foo as (SELECT * FROM hello.world)
            CREATE TABLE egg.spam AS
            SELECT a, b FROM foo`)
                )[0]
            )
        ).toEqual('create');
    });

    test('Simple insert with multiple "with" statements', () => {
        expect(
            getStatementType(
                simpleParse(
                    tokenize(`WITH
                    x AS (SELECT a FROM t),
                    y AS (SELECT a AS b FROM x),
                    z AS (SELECT b AS c FROM y)
                  INSERT c FROM SELECT * FROM z; `)
                )[0]
            )
        ).toEqual('insert');
    });

    // with case with many brackets
    test('Complex with statement', () => {
        expect(
            getStatementType(
                simpleParse(
                    tokenize(`with p as (
            select *
            from (values
                ('a', 1, 1),
                ('b', 2, null),
                ('c', null, 3),
                ('d', null, null)
            ) t1 (letter, val1, val2)
            ), z as (
            select *, val1 IS DISTINCT FROM val2
            from p
            order by letter
            )
            INSERT c FROM SELECT * FROM z;`)
                )[0]
            )
        ).toEqual('insert');
    });

    test('Invalid query with no keyword', () => {
        expect(
            getStatementType(simpleParse(tokenize('selec 1;'))[0])
        ).toBeNull();
    });
});
describe('tokenPatternMatch', () => {
    test('Simple example query contains limit', () => {
        expect(
            tokenPatternMatch(simpleParse(tokenize(simpleQuery))[0], [
                {
                    text: 'limit',
                    type: 'KEYWORD',
                },
                {
                    type: 'NUMBER',
                },
            ])
        ).toBeTruthy();
    });
    test('Simple example query does not contain with', () => {
        expect(
            tokenPatternMatch(simpleParse(tokenize(simpleQuery))[0], [
                {
                    text: 'with',
                    type: 'KEYWORD',
                },
            ])
        ).toBeFalsy();
    });
    test('Simple insert query contains insert', () => {
        expect(
            tokenPatternMatch(
                simpleParse(
                    tokenize(`INSERT INTO abc SELECT * FROM foobar`)
                )[0],
                [
                    {
                        text: 'insert',
                        type: 'KEYWORD',
                    },
                    {
                        text: 'into',
                        type: 'KEYWORD',
                    },
                    {
                        type: 'VARIABLE',
                    },
                ]
            )
        ).toBeTruthy();
    });
    test('Invalid query without keyword', () => {
        expect(
            tokenPatternMatch(simpleParse(tokenize('selec 1;'))[0], [
                {
                    type: 'KEYWORD',
                },
            ])
        ).toBeFalsy();
    });
});
