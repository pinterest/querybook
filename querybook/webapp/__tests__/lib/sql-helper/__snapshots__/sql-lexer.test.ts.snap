// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleParse 1`] = `
Array [
  Array [
    Object {
      "end": 10,
      "line": 1,
      "start": 4,
      "text": "select",
      "type": "KEYWORD",
    },
    Object {
      "end": 17,
      "line": 1,
      "start": 13,
      "text": "from",
      "type": "KEYWORD",
    },
    Object {
      "end": 23,
      "line": 1,
      "start": 18,
      "text": "table",
      "type": "KEYWORD",
    },
    Object {
      "end": 29,
      "line": 1,
      "start": 24,
      "text": "where",
      "type": "KEYWORD",
    },
    Object {
      "end": 36,
      "line": 1,
      "start": 30,
      "text": "fielda",
      "type": "VARIABLE",
    },
    Object {
      "end": 40,
      "line": 1,
      "start": 39,
      "text": "5",
      "type": "NUMBER",
    },
  ],
  Array [
    Object {
      "end": 10,
      "line": 2,
      "start": 4,
      "text": "select",
      "type": "KEYWORD",
    },
    Object {
      "end": 17,
      "line": 2,
      "start": 13,
      "text": "from",
      "type": "KEYWORD",
    },
    Object {
      "end": 23,
      "line": 2,
      "start": 18,
      "text": "table",
      "type": "KEYWORD",
    },
    Object {
      "end": 29,
      "line": 2,
      "start": 24,
      "text": "where",
      "type": "KEYWORD",
    },
    Object {
      "end": 36,
      "line": 2,
      "start": 30,
      "text": "fieldb",
      "type": "VARIABLE",
    },
    Object {
      "bracketIndex": 13,
      "end": 40,
      "line": 2,
      "start": 39,
      "text": "(",
      "type": "BRACKET",
    },
    Object {
      "end": 46,
      "line": 2,
      "start": 40,
      "text": "select",
      "type": "KEYWORD",
    },
    Object {
      "end": 53,
      "line": 2,
      "start": 47,
      "text": "fielda",
      "type": "VARIABLE",
    },
    Object {
      "end": 58,
      "line": 2,
      "start": 54,
      "text": "from",
      "type": "KEYWORD",
    },
    Object {
      "end": 64,
      "line": 2,
      "start": 59,
      "text": "table",
      "type": "KEYWORD",
    },
    Object {
      "end": 70,
      "line": 2,
      "start": 65,
      "text": "where",
      "type": "KEYWORD",
    },
    Object {
      "end": 77,
      "line": 2,
      "start": 71,
      "text": "fieldb",
      "type": "VARIABLE",
    },
    Object {
      "end": 81,
      "line": 2,
      "start": 80,
      "text": "5",
      "type": "NUMBER",
    },
    Object {
      "end": 82,
      "line": 2,
      "start": 81,
      "text": ")",
      "type": "BRACKET",
    },
  ],
]
`;

exports[`simpleParse 2`] = `
Array [
  Array [
    Object {
      "end": 6,
      "line": 0,
      "start": 0,
      "text": "select",
      "type": "KEYWORD",
    },
    Object {
      "end": 13,
      "line": 0,
      "start": 9,
      "text": "from",
      "type": "KEYWORD",
    },
    Object {
      "end": 19,
      "line": 0,
      "start": 14,
      "text": "table",
      "type": "KEYWORD",
    },
    Object {
      "end": 25,
      "line": 0,
      "start": 20,
      "text": "where",
      "type": "KEYWORD",
    },
    Object {
      "end": 33,
      "line": 0,
      "start": 26,
      "text": "extract",
      "type": "KEYWORD",
    },
    Object {
      "bracketIndex": 9,
      "end": 34,
      "line": 0,
      "start": 33,
      "text": "(",
      "type": "BRACKET",
    },
    Object {
      "end": 38,
      "line": 0,
      "start": 34,
      "text": "YEAR",
      "type": "VARIABLE",
    },
    Object {
      "end": 43,
      "line": 0,
      "start": 39,
      "text": "from",
      "type": "KEYWORD",
    },
    Object {
      "end": 50,
      "line": 0,
      "start": 44,
      "text": "fielda",
      "type": "VARIABLE",
    },
    Object {
      "end": 51,
      "line": 0,
      "start": 50,
      "text": ")",
      "type": "BRACKET",
    },
    Object {
      "end": 55,
      "line": 0,
      "start": 54,
      "text": "5",
      "type": "NUMBER",
    },
  ],
]
`;

exports[`tokenize 1`] = `
Array [
  Object {
    "end": 6,
    "line": 0,
    "start": 0,
    "text": "select",
    "type": "KEYWORD",
  },
  Object {
    "end": 8,
    "line": 0,
    "start": 7,
    "text": "*",
    "type": "OPERATOR",
  },
  Object {
    "end": 13,
    "line": 0,
    "start": 9,
    "text": "from",
    "type": "KEYWORD",
  },
  Object {
    "end": 19,
    "line": 0,
    "start": 14,
    "text": "table",
    "type": "KEYWORD",
  },
  Object {
    "end": 25,
    "line": 0,
    "start": 20,
    "text": "where",
    "type": "KEYWORD",
  },
  Object {
    "end": 27,
    "line": 0,
    "start": 26,
    "text": "a",
    "type": "VARIABLE",
  },
  Object {
    "end": 29,
    "line": 0,
    "start": 28,
    "text": "=",
    "type": "OPERATOR",
  },
  Object {
    "end": 31,
    "line": 0,
    "start": 30,
    "text": "5",
    "type": "NUMBER",
  },
  Object {
    "end": 32,
    "line": 0,
    "start": 31,
    "text": ";",
    "type": "SEMI",
  },
]
`;

exports[`tokenize 2`] = `
Array [
  Object {
    "end": 6,
    "line": 1,
    "start": 0,
    "text": "select",
    "type": "KEYWORD",
  },
  Object {
    "end": 6,
    "line": 2,
    "start": 2,
    "text": "aa.a",
    "type": "VARIABLE",
  },
  Object {
    "end": 7,
    "line": 2,
    "start": 6,
    "text": ",",
    "type": "COMMA",
  },
  Object {
    "end": 6,
    "line": 3,
    "start": 2,
    "text": "aa.b",
    "type": "VARIABLE",
  },
  Object {
    "end": 4,
    "line": 4,
    "start": 0,
    "text": "from",
    "type": "KEYWORD",
  },
  Object {
    "end": 9,
    "line": 5,
    "start": 2,
    "text": "table_a",
    "type": "VARIABLE",
  },
  Object {
    "end": 12,
    "line": 5,
    "start": 10,
    "text": "aa",
    "type": "VARIABLE",
  },
  Object {
    "end": 6,
    "line": 6,
    "start": 2,
    "text": "join",
    "type": "KEYWORD",
  },
  Object {
    "end": 14,
    "line": 6,
    "start": 7,
    "text": "table_b",
    "type": "VARIABLE",
  },
  Object {
    "end": 17,
    "line": 6,
    "start": 15,
    "text": "bb",
    "type": "VARIABLE",
  },
  Object {
    "end": 20,
    "line": 6,
    "start": 18,
    "text": "on",
    "type": "KEYWORD",
  },
  Object {
    "end": 25,
    "line": 6,
    "start": 21,
    "text": "aa.a",
    "type": "VARIABLE",
  },
  Object {
    "end": 27,
    "line": 6,
    "start": 26,
    "text": "=",
    "type": "OPERATOR",
  },
  Object {
    "end": 32,
    "line": 6,
    "start": 28,
    "text": "bb.b",
    "type": "VARIABLE",
  },
  Object {
    "end": 5,
    "line": 7,
    "start": 0,
    "text": "where",
    "type": "KEYWORD",
  },
  Object {
    "end": 6,
    "line": 8,
    "start": 2,
    "text": "aa.b",
    "type": "VARIABLE",
  },
  Object {
    "end": 8,
    "line": 8,
    "start": 7,
    "text": "=",
    "type": "OPERATOR",
  },
  Object {
    "end": 10,
    "line": 8,
    "start": 9,
    "text": "3",
    "type": "NUMBER",
  },
  Object {
    "end": 5,
    "line": 9,
    "start": 0,
    "text": "order",
    "type": "KEYWORD",
  },
  Object {
    "end": 8,
    "line": 9,
    "start": 6,
    "text": "by",
    "type": "KEYWORD",
  },
  Object {
    "end": 6,
    "line": 10,
    "start": 2,
    "text": "bb.b",
    "type": "VARIABLE",
  },
  Object {
    "end": 5,
    "line": 11,
    "start": 0,
    "text": "limit",
    "type": "KEYWORD",
  },
  Object {
    "end": 3,
    "line": 12,
    "start": 2,
    "text": "5",
    "type": "NUMBER",
  },
]
`;
